plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -398,6 +393,157 @@ private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProv
         return clientBuilder.build();
     }
 
+    @NotThreadSafe
+    private static class S3ExchangeStorageReader
+            implements ExchangeStorageReader
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(S3ExchangeStorageReader.class).instanceSize();
+
+        private final S3AsyncClient s3AsyncClient;
+        private final Iterator<ExchangeSourceFile> exchangeSourceFiles;
+        private final int partSize;
+        private final int bufferSize;
+
+        private ExchangeSourceFile currentFile;
+        private long fileOffset;
+
+        private byte[] buffer;
+        private int bufferPosition;
+        private int bufferFill;
+
+        private int sliceSize = -1;
+        private ListenableFuture<Void> inProgressReadFuture = immediateVoidFuture();
+
+        public S3ExchangeStorageReader(
+                S3AsyncClient s3AsyncClient,
+                Iterator<ExchangeSourceFile> exchangeSourceFiles,
+                int partSize,
+                int maxPageStorageSize)
+        {
+            this.s3AsyncClient = requireNonNull(s3AsyncClient, "s3AsyncClient is null");
+            this.exchangeSourceFiles = requireNonNull(exchangeSourceFiles, "exchangeSourceFiles is null");
+            this.partSize = partSize;
+            // Make sure buffer can accommodate at least one complete Slice, and keep reads aligned to part boundaries
+            this.bufferSize = maxPageStorageSize + partSize;
+
+            // This class is expected to be used in a single-threaded manner, therefore assignment to non-final fields
+            // doesn't constitute unsafe publication
+            checkArgument(exchangeSourceFiles.hasNext(), "No exchange source file to read from");
+            this.currentFile = exchangeSourceFiles.next();
+            fillBuffer();
+        }
+
+        @Override
+        public Slice read()
+                throws IOException
+        {
+            if (currentFile == null) {
+                return null;
+            }
+
+            try {
+                getFutureValue(inProgressReadFuture);
+            }
+            catch (RuntimeException e) {
+                Throwable cause = e.getCause() == null ? e : e.getCause();
+                throw new IOException(cause);
+            }
+
+            verify(bufferFill - bufferPosition > Integer.BYTES, "Buffer content should contain at least one complete slice");
+            if (sliceSize < 0) {
```

Comment:
I dropped `sliceSize` and extracted a few methods. I think it is more readable this way (I did not test to be fair - so maybe I made a type somewhere):
```java
        @Override
        public Slice read()
                throws IOException
        {
            if (currentFile == null) {
                return null;
            }

            try {
                getFutureValue(inProgressReadFuture);
            }
            catch (RuntimeException e) {
                Throwable cause = e.getCause() == null ? e : e.getCause();
                throw new IOException(cause);
            }

            verify(canReadSliceFromBuffer(), "Buffer content should contain at least one complete slice");
            Slice slice = getSliceFromBuffer();
            if (!canReadSliceFromBuffer()) {
                // we need to read more date to read next slice
                fillBuffer();
            }
            return slice;
        }

        private boolean canReadSliceFromBuffer()
        {
            return bufferFill - bufferPosition > Integer.BYTES && bufferFill - bufferPosition - Integer.BYTES >= pickSliceSizeFromBuffer();
        }

        private int getSliceSizeFromBuffer()
        {
            int sliceSize = pickSliceSizeFromBuffer();
            bufferPosition += Integer.BYTES;
            return sliceSize;
        }

        private int pickSliceSizeFromBuffer()
        {
            return ByteBuffer.wrap(buffer, bufferPosition, Integer.BYTES).order(LITTLE_ENDIAN).getInt();
        }

        private Slice getSliceFromBuffer()
        {
            int sliceSize = getSliceSizeFromBuffer();
            Slice slice = Slices.wrappedBuffer(buffer, bufferPosition, sliceSize);
            bufferPosition += sliceSize;
            return slice;
        }
```

Reply:
I have changed to use `sliceInput`, and the code is much more readable now.